docs/source_rewriter.md

```diff
@@ -0,0 +1,210 @@
+objcopy --redefine-sym foo=foo_from_1 comp1.o
+objcopy --redefine-sym foo=foo_from_2 comp2.o
+
+gcc -fPIC -shared comp1.o -o libcomp1.so -Wl,--wrap=foo_from_1
+gcc -fPIC -shared comp2.o -o libcomp2.so -Wl,--wrap=foo_from_2
+```
+
+and we'll generate the following call gate wrappers
+
+```
+asm(__wrap_foo_from_1:
+      wrpkru pkey3
+      call foo
+      wrpkru pkey1
+      ret);
+
+asm(__wrap_foo_from_2:
+      wrpkru pkey3
+      call foo
+      wrpkru pkey2
+      ret);
+```
+
+
+### Call gates for pointers
+
+Call gates for pointers come in two variants: wrappers for internal and external
+linkage. Again the tool has two phases. The first phase will search for
+functions that have their address taken, making sure to store functions in
+different sets based on their linkage. For internal linkage we need to track the
+files where the function had its address taken. For external linkage we just
+need the function name. It'll also rewrite these function pointer expressions as
+`IA2_FN(name)`.
+
+In the second phase, we'll generate the call gate wrappers. Since a function
+pointer may be freely moved between compartments, the wrapper will always expect
+the default untrusted pkey (the callsite wrappers will change to this pkey
+before calling a pointer). A function has a fixed compartment so the target
+pkeys in wrappers for pointers will also be fixed. Since the caller/target pkeys
+cannot take different values, we can just name the wrappers for the external
+linkage case `__ia2_$FUNCTION_NAME` and write them to the output `.c`. The
+`IA2_FN` macro mentioned above will then take care of prepending `__ia2_` and
+the output header will contain declarations for the wrappers.
+
+For internal linkage (i.e. static) functions that have their address taken, the
+same caller/target pkey logic applies so we can also name their wrappers by
+prepending `__ia2_`. However since they call `static` functions they must be
+defined in the same translation unit as the `static` function. To avoid
+cluttering up the input source files, we'll define the following macro in the
+output header for each of these functions and invoke them in the corresponding
+input sources.
+
+```
+#define IA2_DEFINE_STATIC_WRAPPER_FN_NAME \
+    asm(__ia2_fn_name:                    \
+          wrpkru pkey                     \
+          call fn_name                    \
+          wrpkru 0                        \
+          ret);
+```
+
+This wrapper will be identical to the wrappers for external linkage except that
+it will not have `.global __ia2_fn_name` to make the wrapper static. This will
+make it easier to support the case where static functions in different files
+with the same name have their address taken. TODO: We write the
+`IA2_DEFINE_STATIC_WRAPPER_*` functions to one output header so we'd need a way
+to distinguish different files.
+
+### Call gates for indirect callsites
+
+We also need call gates for indirect callsites. Each compartment might not
```

Comment:
Do we rewrite every indirect callsite? Is there a way to opt out for hot callsites known to only call functions internal to its own compartment?